l10n_de_tax_statement/__manifest__.py

```diff
@@ -0,0 +1,26 @@
+# Copyright 2019 Onestein (<https://www.onestein.eu>)
+# Copyright 2019 BIG-Consulting GmbH(<http://www.openbig.org>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+{
+    'name': 'German VAT Statement',
+    'version': '11.0.1.1.0',
+    'category': 'Localization',
+    'license': 'AGPL-3',
+    'author': 'OpenBIG.org, Onestein, Odoo Community Association (OCA)',
+    'website': 'https://github.com/openbig/l10n-germany',
+    'depends': [
+        'account_tax_balance',
```

Comment:
I think that the dependency on module `l10n_de` should be not here. The reason is that this module was meant to be compatible with every German localization module (eg.: any custom localization module for Germany).
The logic implemented in the [configuration wizard](https://github.com/OCA/l10n-germany/blob/937ccd80bd174876e6f8534c99743aaa1dc6ff7e/l10n_de_tax_statement/wizard/l10n_de_tax_statement_config_wizard.py) is adaptive: if `l10n_de` is already installed, when installing this module the tags are recognized and included in the configuration.

Reply:
In that way the adaptive implementation isn't working because the tests are throwing errors because of the missing xmlids. In this case you should check if the `l10n_de.tag_de_xx` are present in the lines 114+ in the wizard.

Which different german localization modules do you mean?

Reply:
Thanks, indeed i was thinking about that too, because i in the adopted tests i have referenced this tag. Hopefully that will fix the testing...

Reply:
@fkantelberg This module was initially made for the Netherlands localization. In the Netherlands the Chart of Accounts (COA) is free, every company could adopt its own COA, not necessarily the standard one provided by Odoo SA. Not sure is the same case for Germany.

Reply:
Ah. Can we generalize this module and offer a glue module if the l10n_de is installed with the wizard configured properly?

Reply:
@fkantelberg when this logic will be fixed and working properly, not any glue module will be necessary (that was exactly the reason of implementing this logic).

Reply:
On my local test system i could install the module and it was working fine for both scenarios (without the l10n_de dependency in the module __manifest__.py):
a. German company with accounting plan skr03.
b. German company with accounting plan skr04.

Anyway finally a german company will ending either with l10n_de + l10n_de_skr03 or l10n_de + l10n_de_skr04 if they will decide to add on top this tax declaration module (l10n_de_tax_statement).

Theoretically there could be a scenario "c." : a german company doesn't want to install l10n_de and / or l10n_de_skr03 or l10n_de_skr04 neither. 

@astirpe I assume that is the scenario you mean ?

Reply:
@tv-openbig Yes, is theoretical scenario "c" possible in a real case? If yes, then my previous comments would be still valid.

Reply:
So the initial idea for this module was to make it independent by the chosen localization module. But in case you already had the `l10n_de` installed (that would be often the case), this module can load the proper tags in the configuration automatically.

Reply:
@astripe Then I think it was a misunderstanding but to make this module independent you should just delete everything which depends on l10n_de and put the dependent stuff into a new glue module  which will automatically installed when l10n_de and this one are installed to configure for the l10n_de case. This would mean that the glue module won't be installed in the scenario c but in a and b where it can set the configuration for the german localization.

Reply:
> @astirpe I think this is for you. poor choice of user name on my part I guess.

Reply:
@fkantelberg
When I say "independent" I mean: not containing `l10n_de` in the dependencies of the `__manifest__.py` file.
When I say "adaptive" I mean: the configuration is automatically set in case `l10n_de` is already installed, which is a common case.

The glue logic would be fine for scenario "c". But when we talk about the standard Odoo localization modules (scenarios "a" and "b"), the glue logic can be avoided IMO. So the end users would be happy to just install one module instead of two or more.

I'm not very into the German community, so for me doesn't really matter which logic you choose. I just wanted to help to understand this module better. If you are happy with two modules instead of one for the case "standard German localization + OCA tax statement", I'm not going to block it :wink:

Reply:
@astirpe There was the misunderstanding. :+1: 
You want to minimize the number of module because the user only have to install 1. Same here. You can achieve it by setting the `auto_install` to true in the manifest. Odoo will install it if the dependencies are fulfilled.

Reply:
@fkantelberg Some customer could forget to download the auto_install modules, or just could not notice that those modules exist. And there are many other situations where you would prefer just one plug-and-play module, instead of handling multiple simple modules.
If you are worried about the code readability, still is possible to write a good code without splitting it in multiple modules.
But I'm not going to insist anymore. :smile:

Reply:
@astirpe @fkantelberg The module works fine. It could be installed on top of the two only existing modules without any issues. For the case a user wants to install that module without any l10n_de module installed i assume it would also work if you configure  the tax statement in advance by hand before you would be able to get a proper report. 

As a functional consultant i can basically understand the code. It was enough to adopt it. For sure it is more important to have background knowledge of german accounts, taxes, tax declaration as coding knowledge for that kind of module. For financial accountants as end users it is self-explainable.

The mixing of german terms and english terms is non-avoidable or better it doesn't make sense to back-translate the german terms in english.

I would be glad if we could push somehow that module through the testing as i wanted to proceed to create next PR's for german balance sheets / p&l report. If you could help to push that module through the remaining technical barriers it would be really great.

Reply:
@astirpe @fkantelberg I have now checked locally. 

If you select germany as country on database installation and install at first the dependency of the module l10n_de_tax_statement (=account_tax_balance) as a first step nothing happens as the dependency of account_tax_balance is imho wrong or outdated. Nothing happens means, i cannot see accounting menu, which is definitely a requirement if somebody wants to use the tax declaration module or wants to proceed testing on runbot.

I solved that so far by hand via installation of account_invoicing module. As a consequence of account_tax_balance dependency issue i would eventually propose to add that module as another dependecy in our module until that is fixed .

If in fact both comes together 1.) german company and 2.) account_invoicing module the l10n_de module as well l10n_de_skr03 and l10n_de_skr04 will be installed automatically. Now a user can  switch between both accounting plans until he creates first account moves. As a consequence for our l10n_de_tax_statement module the tax reporting could be used out of the box in both cases.

If i want to force scenario c.) described as „a german company wants to install some other non l10n_de related account chart“  i have to uninstall at first all official l10n_de modules if i marked the companies country as germany. If i install now f.e. l10n_nl or any other non l10n_de account chart indeed i get exactlly the error mentioned by travis automated testing:
  File "/odoo/odoo-server/odoo/addons/base/ir/ir_model.py", line 1324, in xmlid_lookup
    raise ValueError('External ID not found in the system: %s' % xmlid)
ValueError: External ID not found in the system: l10n_de.tag_de_41

I didn’t get that error anymore if l10n_de is a requirement of our module, even if i really want to use l10n_nl account chart in combination with l10n_de_tax_statement. 

Therefore i would propose to follow the proposal of @fkantelberg to add the l10n_de dependency.

The only drawback in that theoretical case would be that the tax statement configuration would need a reconfiguration to a.) correct the mappings to the NL tax tags or to adopt the NL tax -> DE tax tag mapping. I assume there is no other way to do that anyway. 

A future improvement could be that in a case if there is another accounting plan installed which doesn’t relate to  l10n_de to clean up the tax statement configuration wizard in order to allow a by hand tax statement configuration from scratch without any pre-filled tax statement configuration. I assume exactly that is covered by l10n_nl_vat_statement module in order to cover all theoretical cases of a account chart installation. 

Finally my proposal is to put that in the # TO DO section of the modules description, except anyone could help to solve that 5% case. Not to forget a reconfiguration could be done, so this  limited functionality is not an error. I assume if everybody agrees we should identify the lines we may uncomment in the source code. 

@astirpe Could you point to the lines to uncomment under that circumstances in l10n_de_tax_statement_config_wizard ?

Reply:
@tv-openbig 

The following error:

```
File "/odoo/odoo-server/odoo/addons/base/ir/ir_model.py", line 1324, in xmlid_lookup
raise ValueError('External ID not found in the system: %s' % xmlid)
ValueError: External ID not found in the system: l10n_de.tag_de_41
```
is caused by the wrong indentation issue, already mentioned in our previous comments.

Once you fix it, the dependency `l10n_de` could be safely removed.

Reply:
The indentation issue is causing this travis error:

```
l10n_de_tax_statement/wizard/l10n_de_tax_statement_config_wizard.py:111: [W0101(unreachable), VatStatementConfigWizard.default_get] Unreachable code
```